lib/posix/pthread_barrier.c

```diff
@@ -4,31 +4,177 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#include "posix_internal.h"
+
+#include <zephyr/init.h>
 #include <zephyr/kernel.h>
 #include <zephyr/posix/pthread.h>
-#include <ksched.h>
-#include <zephyr/wait_q.h>
+#include <zephyr/sys/bitarray.h>
+
+struct posix_barrier {
+	struct k_mutex mutex;
+	struct k_condvar cond;
+	uint32_t max;
+	uint32_t count;
+};
+
+static struct posix_barrier posix_barrier_pool[CONFIG_MAX_PTHREAD_BARRIER_COUNT];
+SYS_BITARRAY_DEFINE_STATIC(posix_barrier_bitarray, CONFIG_MAX_PTHREAD_BARRIER_COUNT);
+
+/*
+ * We reserve the MSB to mark a pthread_barrier_t as initialized (from the
+ * perspective of the application). With a linear space, this means that
+ * the theoretical pthread_barrier_t range is [0,2147483647].
+ */
+BUILD_ASSERT(CONFIG_MAX_PTHREAD_BARRIER_COUNT < PTHREAD_OBJ_MASK_INIT,
+	     "CONFIG_MAX_PTHREAD_BARRIER_COUNT is too high");
 
-extern struct k_spinlock z_pthread_spinlock;
+static inline size_t posix_barrier_to_offset(struct posix_barrier *bar)
+{
+	return bar - posix_barrier_pool;
+}
+
+static inline size_t to_posix_barrier_idx(pthread_barrier_t b)
+{
+	return mark_pthread_obj_uninitialized(b);
+}
+
+struct posix_barrier *get_posix_barrier(pthread_barrier_t b)
+{
+	int actually_initialized;
+	size_t bit = to_posix_barrier_idx(b);
+
+	/* if the provided barrier does not claim to be initialized, its invalid */
+	if (!is_pthread_obj_initialized(b)) {
+		return NULL;
+	}
+
+	/* Mask off the MSB to get the actual bit index */
+	if (sys_bitarray_test_bit(&posix_barrier_bitarray, bit, &actually_initialized) < 0) {
+		return NULL;
+	}
+
+	if (actually_initialized == 0) {
+		/* The barrier claims to be initialized but is actually not */
+		return NULL;
+	}
+
+	return &posix_barrier_pool[bit];
+}
 
 int pthread_barrier_wait(pthread_barrier_t *b)
 {
-	k_spinlock_key_t key = k_spin_lock(&z_pthread_spinlock);
-	int ret = 0;
+	int ret;
+	int err;
+	pthread_barrier_t bb = *b;
+	struct posix_barrier *bar;
+
+	bar = get_posix_barrier(bb);
+	if (bar == NULL) {
+		return EINVAL;
+	}
 
-	b->count++;
+	err = k_mutex_lock(&bar->mutex, K_FOREVER);
+	if (err < 0) {
+		return -err;
+	}
+	__ASSERT_NO_MSG(err == 0);
```

Comment:
I think originally, I had it as a direct assert,  without the conditional. That would be my preference. I would have to go back to my laptop to see what the context is as to whether that is ok in this situation.

Reply:
The assert works

Reply:
> The assert works

Probably I don't understand your comment. What do you mean by 'work'? Do you mean that it's not triggered? 

My concern here is that we **can't** possibly trigger it. `k_mutex_lock` as zephyr API return either 0 or negative error code. In case of error we return before the assert, so assert will always be executed with `err = 0`.